library/core/src/panic.rs

```diff
@@ -139,27 +139,30 @@ pub macro unreachable_2021 {
     ),
 }
 
-/// Asserts that a boolean expression is `true`, and perform a non-unwinding panic otherwise.
+/// Just like `assert_unsafe_precondition!` the defining features of this macro are that its
+/// checks are enabled when they are monomorphized with debug assertions enabled, and upon failure
+/// a non-unwinding panic is launched so that failures cannot compromise unwind safety.
 ///
-/// This macro is similar to `debug_assert!`, but is intended to be used in code that should not
-/// unwind. For example, checks in `_unchecked` functions that are intended for debugging but should
-/// not compromise unwind safety.
+/// Unlike `assert_unsafe_precondition!` the condition being checked here is not put in an outlined
+/// function. If the check compiles to a lot of IR, this can cause code bloat if the check is
+/// monomorphized many times. But it also means that the checks from this macro can be deduplicated
+/// or otherwise optimized out.
```

Comment:
It's referring to `assert_unsafe_precondition` a lot, but I think one important distinction is that this isn't unsound to call this with `false` since it doesn't use `const_eval_select`.